Added support for upgrading SPFx@1.0.2 to 1.1.0 projects solving #537

baywet opened the pull request:
@waldekmastykarz last one, and this one is a bit more hairy.
First I had to start off #500 because I didn't want to have to generate the test files and end up with conflicts. I merged dev into this branch which is initially a fork of the branch supporting PR 500 in your repo.
Second, the question about the react dependencies stands like in #536 
Third, there are two things I left on the side for now because I wanted to discuss with you before proceeding:
- .vscode/settings.json was introduced in 1.1.0, so it's missing from older projects. We should probably tell people to add it, but should I copy the full source in the resolution or tell them to leverage a wget with a permalink pointing somewhere?
- the webpart manifest version switched from 0.0.1 to *, I'm not decided whether we should tell people to change that systematically, what do you think?

waldekmastykarz commented:
We should provide the full contents of `.vscode/settings.json` as specified by v1.1.0. We should suggest using * but as an optional change. That way, developers can decide themselves whether they want to use the automatic versioning or stick to manual.

baywet commented:
I'll try to implement the vscode settings and the version information ASAP.
You probably noticed that this PR's associated build keeps failing. This is because the path for a file is different when running unit tests locally than with circle ci.
I couldn't figure it out after a few trials, so rather than polluting the history with lots of trials and errors I figured you probably could help with that?

waldekmastykarz commented:
Absolutely! I'll have a look. Btw, when creating rules, please follow the naming convention of ID_AREA_RuleName like we do in other rules

baywet commented:
I tried to keep following that, which one is faulting from your perspective?

waldekmastykarz commented:
The FileRemoveRule doesn't need to be implemented as an abstract rule since all it does is to check for a presence of the file specified in the rule's constructor. You can implement it as a non-abstract class and then in the upgrade-1.0.2.ts file, when creating the instance of the rule, specify the file name in the constructor. If there are multiple files that need to be removed, you can create multiple instances of the rule.

waldekmastykarz commented:
Regarding the test, you should mock `fs.existsSync` rather than trying to construct a real path. As you've noticed, that can't be done reliably across all the different machines. If you need an example how to do the mock, check out https://github.com/baywet/office365-cli/blob/7716e7c1b611f8b1ed7f541358d826a13632c73a/src/appInsights.spec.ts#L13. Don't forget to include https://github.com/baywet/office365-cli/blob/7716e7c1b611f8b1ed7f541358d826a13632c73a/src/appInsights.spec.ts#L7-L10 (excl. clearing the require cache) or other tests that depend on fs.existsSync will fail.

baywet commented:
@waldekmastykarz thanks for all the guidance!
Implemented the latest changes we talked about:
- added vscode settings recommendation rule
- removed unecessary abstract on file remove and deleted unnecessary associated rules
- added webpart manifest rule
- fixed unit tests for file remove rule

With all that it should be good for merging and testing. Unless you have more feedback.

waldekmastykarz commented:
Thanks! It seems like the code coverage dropped significantly. Are you missing some tests by any chance?

baywet commented:
I noticed that too on this build https://coveralls.io/builds/18553115 (previous commit) as I had a "New Missed Lines in Diff" section and improved my test cases to augment coverage. Now on this build (https://coveralls.io/builds/18553171) it looks like the coverage improved a bit (from 98.29 to 98.31) but I can't see any of "my code" being faulting unless I'm ready it wrong.

waldekmastykarz commented:
The dev branch to which you're submitting the PR has coverage of 100% so it could be that either your fork is out of sync or there is something missing something after all.

baywet commented:
I just double checked my sync with the dev branch, looks ok so far. #540 and #541 look fin in terms of coverage. The only difference for this one is I started on top of your PR. I'm going to create a new branch (based of dev) and cherry pick my commits there, see if it makes any difference and I somehow messed up my merges.

waldekmastykarz commented:
Sometimes commits get tangled and merges miss something. Hopefully we'll be able to sort it out quickly 👍

baywet commented:
closing because of merging and coverage issues, moved to #551
